crypto/multisig/src/lib.rs

```diff
@@ -101,10 +133,17 @@ impl<P: Default + PublicKey + Message> SignerSet<P> {
     }
 
     /// Check if this signer set is valid.
-    /// A signer set is considered valid if it has a threshold of at least one,
-    /// and the number of signers is greater than or equal to the threshold.
+    /// A signer set is considered valid if:
+    /// - All nested signer sets underneath it are also valid
+    /// - It has a threshold of at least one
+    /// - The number of signers is greater than or equal to the threshold.
     pub fn is_valid(&self) -> bool {
-        0 < self.threshold && self.threshold as usize <= self.signers.len()
+        // All nested sets must be valid
+        if !self.multi_signers.iter().all(|s| s.is_valid()) {
+            return false;
+        }
+
+        0 < self.threshold && self.threshold as usize <= self.num_signers()
```

Comment:
Optional/Equivalent: Merge into one ANDed expression
```suggestion
        self.multi_signers.iter().all(|s| s.is_valid()) &&
            0 < self.threshold && self.threshold as usize <= self.num_signers()
```